Integrate Chucker for network troubleshooting

oguzkocer opened the pull request:
## Summary

Integrates [Chucker](https://github.com/ChuckerTeam/chucker) to allow users to inspect HTTP traffic for troubleshooting network issues when contacting support.

**Key points:**
- Production-ready but disabled by default - users opt-in via toggle in Help/Support screen
- Privacy-focused: logs stay on-device, sensitive headers redacted (`Authorization`, `Cookie`, `Set-Cookie`, `X-WP-Nonce`)
- Configurable retention periods (1 hour, 1 day, 1 week, until cleared)
- Behind `NETWORK_DEBUGGING` experimental feature flag
- Covers OkHttp clients, wordpress-rs clients, and GutenbergKit WebView requests

## Changes

- Add `TrackNetworkRequestsInterceptor` wrapper around Chucker in FluxC
- Add toggle UI to Help screen and new Support screen (Compose)
- Integrate interceptor into all network paths:
  - Main OkHttpClient (via Dagger multibindings)
  - `WpApiClientProvider`, `WPcomLoginClient`, `WpLoginClient`
  - GutenbergKit via `GutenbergKitNetworkLogger` (replays JS fetch through OkHttp)
- Store preferences at device level (`UndeletablePrefKey`) to support login flow debugging

## Test Plan

### 1. Feature toggle flow (see GIF 1)

Verify the following by following along with the GIF:

- [x] Feature toggle hidden when `NETWORK_DEBUGGING` experimental flag is disabled
- [x] Toggle appears in both Help screen and Support screen when flag is enabled
- [x] Enabling tracking shows retention period picker dialog
- [x] Disabling tracking shows confirmation dialog
- [x] Cannot disable experimental flag while tracking is active (shows error dialog)

### 2. Network requests captured (see GIF 2)

Enable `GUTENBERG_KIT` and `NETWORK_DEBUGGING` experimental flags, then:

1. Open My Site → More → Users
2. Open My Site → More → Subscribers
3. Open the block editor (create/edit a post)
4. Go to Help/Support → View Network Requests

Verify the following requests are captured:

- [x] **FluxC**: Search for `users` → find `GET /rest/v1.1/sites/{site_id}/users`
- [x] **wordpress-rs**: Search for `subscribers` → find `GET /wpcom/v2/sites/{site_id}/subscribers`
- [x] **GutenbergKit**: Search for `block-editor` → find `GET /wp-block-editor/v1/sites/{site_id}/settings`

[Feature toggle flow](https://github.com/user-attachments/assets/dd05d172-0301-400f-87ad-332e8b0af030)

[Network requests captured](https://github.com/user-attachments/assets/86d03e48-6d51-495f-a3c7-24ce83dd7ca4)

---

Part of CMM-998

oguzkocer committed:
Add Chucker integration research documentation

Document the plan for integrating Chucker HTTP inspector into the app
for production use (disabled by default, user opt-in for troubleshooting).

Research findings:
- 7 network paths identified, 5 are OkHttp-based (interceptable)
- OkHttpClientModule is the main path covering ~90% of traffic
- 2 HttpURLConnection usages cannot be intercepted (edge cases)
- Wrapper interceptor approach recommended for runtime enable/disable

oguzkocer committed:
Add track network requests interceptor foundation for Chucker integration

Add an OkHttp interceptor that will serve as the foundation for HTTP
traffic inspection via Chucker. The interceptor is disabled by default
and can be toggled by users in the Help screen for troubleshooting.

Changes:
- Add `TrackNetworkRequestsInterceptor` in FluxC with preference interface
- Add `TrackNetworkRequestsModule` to inject interceptor into OkHttpClient
- Add preference `IS_TRACK_NETWORK_REQUESTS_ENABLED` in `AppPrefs`
- Add toggle UI in Help screen below "Contact email"

oguzkocer committed:
Integrate Chucker library into TrackNetworkRequestsInterceptor

Add Chucker v4.2.0 as the HTTP traffic inspection backend for the
Track Network Requests feature. The interceptor now delegates to
ChuckerInterceptor when the feature is enabled.

Configuration:
- 1-hour retention period
- Sensitive headers redacted (Authorization, Cookie, Set-Cookie, X-WP-Nonce)
- Notifications disabled (custom UI entry point to be added)
- Max content length: 250KB

Changes:
- Add Chucker dependency to libs.versions.toml and build.gradle files
- Update TrackNetworkRequestsInterceptor to create and use ChuckerInterceptor
- Simplify TrackNetworkRequestsModule (Chucker is now an implementation detail)

oguzkocer committed:
Add View Network Requests button to launch Chucker UI

Add a button in the Help screen that launches Chucker's traffic
inspector UI. The button is only visible when Track Network Requests
is enabled.

Changes:
- Add `viewNetworkRequestsButton` to help_activity.xml
- Add `view_network_requests` string resource
- Update `setupTrackNetworkRequestsToggle()` to show/hide button and handle click

oguzkocer committed:
Add configurable retention period and dialog-based UX for network tracking

Implement retention period selection when enabling Track Network Requests,
with dialog-based enable/disable flow for better user communication.

Features:
- Retention period options: 1 Hour, 1 Day, 1 Week, Until Cleared
- Enable dialog with privacy explanation and retention picker
- Disable dialog with guidance on clearing logs manually
- "View Network Requests" now shows current retention period

Technical changes:
- Add `NetworkRequestsRetentionPeriod` enum with stable int values for persistence
- Interceptor lazily creates and recreates ChuckerInterceptor when retention changes
- Add `dialog_title_with_message.xml` layout for dialogs with description text
- Use `setSwitchCheckedSilently()` to prevent listener loops on cancel

Changes:
- Add retention period preference to `AppPrefs` and `AppPrefsWrapper`
- Update `TrackNetworkRequestsPreference` interface with `getRetentionPeriod()`
- Add enable/disable dialogs to `HelpActivity`
- Update `help_activity.xml` with container showing retention info
- Add string resources for retention options and dialog messages

oguzkocer committed:
Add login request tracking and make preferences device-level

Enable network request tracking for OAuth login flows by adding
interceptors to WPcomLoginClient. Move tracking preferences to
UndeletablePrefKey so they persist across logout/login cycles,
allowing users to troubleshoot login issues.

Changes:
- Add interceptors to WPcomLoginClient for OAuth token exchange
- Move IS_TRACK_NETWORK_REQUESTS_ENABLED to UndeletablePrefKey
- Move TRACK_NETWORK_REQUESTS_RETENTION_PERIOD to UndeletablePrefKey
- Add documentation for Chucker export API options

oguzkocer committed:
Add GutenbergKit network request logging to Chucker

Integrates GutenbergKit's WebView network requests with Chucker by
replaying them through OkHttp. This allows editor network traffic
to appear alongside native app requests in the unified log.

Changes:
- Add `GutenbergKitNetworkLogger` to replay JS fetch requests through OkHttp
- Add `isNetworkLoggingEnabled` to `FeatureConfig` and `EditorConfiguration`
- Wire up `NetworkRequestListener` in `GutenbergKitActivity`
- Update GutenbergKit to PR #238 branch with network logging support

oguzkocer committed:
Remove debug logs from GutenbergKit network logging

Remove verbose debug logging used during development.
Keep error logging for diagnosing failures.

oguzkocer committed:
Add experimental feature flag for Network Debugging

Hide the Track Network Requests section in Help behind an experimental
feature flag. When the flag is disabled, the section is completely
hidden. When trying to disable the flag while tracking is enabled,
show an error dialog instructing the user to disable tracking first.

Changes:
- Add NETWORK_DEBUGGING to ExperimentalFeatures.Feature enum
- Add validation in ExperimentalFeaturesViewModel to prevent disabling
  the flag while tracking is active
- Add NetworkDebuggingErrorDialog composable for the error state
- Update HelpActivity to check the feature flag before showing section
- Wrap track network requests UI in a single container for cleaner
  visibility management

oguzkocer committed:
Fix detekt issues in Chucker integration code

Add suppress annotations for:
- LongMethod in ExperimentalFeaturesActivity.onCreate (Compose UI)
- TooGenericExceptionCaught in GutenbergKitNetworkLogger (intentional)
- MagicNumber in GutenbergKitNetworkLogger (HTTP status codes)

oguzkocer committed:
Add network debugging to new Support screen

Duplicate the network tracking functionality from HelpActivity to the
new Compose-based SupportActivity (behind MODERN_SUPPORT flag).

Also update help_activity.xml to show retention info under the toggle
instead of under the view button, for consistency between both UIs.

Changes:
- Add NetworkTrackingState and DialogEvent to SupportViewModel
- Add NetworkTrackingToggleItem composable to SupportScreen
- Add dialog handling and Chucker navigation to SupportActivity
- Restructure help_activity.xml for consistent retention info placement

oguzkocer committed:
Integrate Chucker interceptor into wordpress-rs clients

Wire `TrackNetworkRequestsInterceptor` into all wordpress-rs based
network clients so their requests are captured when network debugging
is enabled.

Changes:
- `WpApiClientProvider`: Accept interceptors via Dagger `@Named` set
- `ApplicationModule`: Pass interceptor to `WpLoginClient`
- `DataViewViewModel`: Inject interceptor for `WpComApiClient`
- `JetpackConnectionHelper`: Inject interceptor for connection client
- `AddSubscribersViewModel`: Inject interceptor for subscriber API
- Child ViewModels: Pass interceptor to parent `DataViewViewModel`

oguzkocer committed:
Update GutenbergKit to v0.11.0 and adapt to API changes

GutenbergKit v0.11.0 renamed `NetworkRequest` to `RecordedNetworkRequest`
and added a `statusText` field, allowing us to remove our custom HTTP
status code mapping.

Changes:
- Update GutenbergKit version to v0.11.0
- Rename NetworkRequest to RecordedNetworkRequest
- Use statusText from RecordedNetworkRequest instead of toStatusMessage()

oguzkocer committed:
Use OkHttpClientQualifiers.INTERCEPTORS constant for @Named annotations

Replace hardcoded "interceptors" strings with the constant to prevent
typos and enable easier refactoring.

oguzkocer committed:
Remove research documentation files

oguzkocer committed:
Fix lint warnings and update tests for new constructor parameters

- Add @NonNull annotation to ApplicationModule parameter
- Add @SuppressLint("InflateParams") for AlertDialog custom title inflation
- Update SupportViewModelTest with appPrefsWrapper and experimentalFeatures
- Update DataViewViewModelTest with trackNetworkRequestsInterceptor
- Update ExperimentalFeaturesViewModelTest with appPrefsWrapper
- Update TermsViewModelTest with trackNetworkRequestsInterceptor

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/modules/ApplicationModule.java:
```diff
@@ -153,7 +154,9 @@
-    public static WpLoginClient provideWpLoginClient() {
-        return new WpLoginClient(Collections.emptyList());
+    public static WpLoginClient provideWpLoginClient(
+            TrackNetworkRequestsInterceptor trackNetworkRequestsInterceptor
```
Should be addressed in 333d16b4ee8649d6dd462c4282d6d8a8f3c88a2d

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/support/main/ui/SupportActivity.kt:
```diff
@@ -81,12 +113,69 @@
+        val displayNames = periods.map { getRetentionPeriodDisplayString(it) }.toTypedArray()
+        var selectedIndex = periods.indexOf(currentPeriod)
+
+        val titleView = layoutInflater.inflate(R.layout.dialog_title_with_message, null).apply {
```
This is intentional - when inflating a custom title view for `AlertDialog.setCustomTitle()`, we don't have access to the dialog's internal parent ViewGroup at inflation time. Passing `null` is the standard approach for this case, suppressed with `@SuppressLint("InflateParams")`. This pattern is used elsewhere in the codebase (e.g., `NotificationsListFragment`, `ThemeBrowserFragment`).

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/HelpActivity.kt:
```diff
@@ -144,6 +155,120 @@
+        var selectedIndex = periods.indexOf(currentPeriod)
+
+        // Custom title view with title + description (setMessage conflicts with setSingleChoiceItems)
+        val titleView = layoutInflater.inflate(R.layout.dialog_title_with_message, null).apply {
```
This is intentional - when inflating a custom title view for `AlertDialog.setCustomTitle()`, we don't have access to the dialog's internal parent ViewGroup at inflation time. Passing `null` is the standard approach for this case, suppressed with `@SuppressLint("InflateParams")`. This pattern is used elsewhere in the codebase (e.g., `NotificationsListFragment`, `ThemeBrowserFragment`).

adalpari commented on WordPress/src/main/java/org/wordpress/android/support/main/ui/SupportActivity.kt:
```diff
@@ -55,20 +63,45 @@ class SupportActivity : AppCompatActivity() {
+            NetworkRequestsRetentionPeriod.FOREVER ->
+                getString(R.string.network_requests_retention_until_cleared)
+        }
+    }
```
⛏️ Very minor readability nitpick

```suggestion
private fun getRetentionPeriodDisplayString(period: NetworkRequestsRetentionPeriod): String {
    return getString(getRetentionPeriodStringRes(period))
}

private fun getRetentionPeriodStringRes(period: NetworkRequestsRetentionPeriod): Int {
    return when (period) {
        NetworkRequestsRetentionPeriod.ONE_HOUR -> R.string.network_requests_retention_one_hour
        NetworkRequestsRetentionPeriod.ONE_DAY -> R.string.network_requests_retention_one_day
        NetworkRequestsRetentionPeriod.ONE_WEEK -> R.string.network_requests_retention_one_week
        NetworkRequestsRetentionPeriod.FOREVER -> R.string.network_requests_retention_until_cleared
    }
}
```

adalpari commented on WordPress/src/main/java/org/wordpress/android/support/main/ui/SupportActivity.kt:
```diff
@@ -81,12 +114,70 @@ class SupportActivity : AppCompatActivity() {
+        var selectedIndex = periods.indexOf(currentPeriod)
+
+        @SuppressLint("InflateParams") // Parent is null because AlertDialog attaches it internally
+        val titleView = layoutInflater.inflate(R.layout.dialog_title_with_message, null).apply {
```
❓ Feels a bit odd to inflate a layout and a dialog inside a Jetpack  Compose activity. What about using a state instead of an event, and a composable AlertDialog?

Note: If we need to unblock this PR, I'm happy to approve it as it is, and work on the changes later.

oguzkocer replied:
Should be addressed in 6505f42c5970d75809fdddcb7073474f84e42c73, let me know what you think!

adalpari commented on WordPress/src/test/java/org/wordpress/android/support/main/ui/SupportViewModelTest.kt:
```diff
@@ -32,7 +40,9 @@ class SupportViewModelTest : BaseUnitTest() {
+            appPrefsWrapper = appPrefsWrapper,
+            experimentalFeatures = experimentalFeatures
         )
     }
```
⛏️ It would be great to add some simple tests for the new functionality in the VM

oguzkocer replied:
Should be addressed in f20caccae04f496ac0200b7f407a8ce035f35c61

Note that I followed the existing patterns in this test file by duplicating the account setup mocks. As a future improvement, we could extract these into shared helpers (e.g., `setupLoggedInAccountMocks()` / `setupLoggedOutAccountMocks()`) to improve readability across all init tests.

adalpari commented on WordPress/src/main/java/org/wordpress/android/support/main/ui/SupportViewModel.kt:
```diff
@@ -105,4 +147,43 @@ class SupportViewModel @Inject constructor(
+    }
+
+    fun onEnableTrackingConfirmed(period: NetworkRequestsRetentionPeriod) {
+        appPrefsWrapper.trackNetworkRequestsRetentionPeriod = period.value
```
I suggest using `viewModelScope` in these functions as well, just to be sure we are not blocking the UI thread

oguzkocer replied:
Could you clarify which operation you're concerned might block the UI thread? The SharedPreferences writes use `apply()` (async), and the StateFlow updates are in-memory. Happy to wrap in `viewModelScope` if there's a specific concern I'm missing\!

oguzkocer committed:
Refactor retention period display string into two methods  Split getRetentionPeriodDisplayString into separate methods for better readability and reusability: - getRetentionPeriodStringRes: returns the string resource ID - getRetentionPeriodDisplayString: resolves it to a String  Changes: - Extract resource ID mapping to dedicated method - Simplify the display string method to delegate

Co-authored-by: Adalberto Plaza <adalpari@gmail.com>

oguzkocer committed:
Migrate network tracking dialogs to Compose

Replace View-based AlertDialogs with Compose AlertDialogs in SupportScreen.
This addresses PR feedback about using Compose dialogs in a Compose activity.

Changes:
- Add reusable SingleChoiceAlertDialog component
- Convert DialogEvent (SharedFlow) to DialogState (StateFlow) in ViewModel
- Add EnableTrackingDialog and DisableTrackingDialog composables
- Remove XML layout inflation and View-based dialog code from Activity

oguzkocer committed:
Add tests for network tracking functionality in SupportViewModel

Cover the new network tracking dialog state management with tests for:
- init() network debugging state based on feature flag and preferences
- onNetworkTrackingToggle showing enable/disable dialogs
- onEnableTrackingConfirmed/onDisableTrackingConfirmed state updates
- onDialogDismissed without side effects
- onRetentionPeriodSelected updating dialog state

Changes:
- Add 10 new tests in SupportViewModelTest
- Move init-related tests to existing init() region
- Create new region for dialog interaction tests

adalpari reviewed (approved):
LGTM and works as expected! Thanks for jumping into the network tracker!
🚢

dcalhoun commented on gradle/libs.versions.toml:
```diff
@@ -73,7 +74,7 @@ google-play-services-auth = '20.4.1'
 gravatar = '2.5.0'
 greenrobot-eventbus = '3.3.1'
-gutenberg-kit = 'v0.10.2'
+gutenberg-kit = 'v0.11.0'
```
I discovered a race condition that needs to be addressed in GutenbergKit. The network activity in GutenbergKit is currently not always logged. I'll open a fix and publish a release.

dcalhoun replied:
@oguzkocer I opened https://github.com/wordpress-mobile/GutenbergKit/pull/243 for review. Once that merges, I'll publish a new GutenbergKit release for integrating into this branch.

oguzkocer replied:
Thanks David! I'll take a look.

dcalhoun replied:
GutenbergKit [v0.11.1](https://github.com/wordpress-mobile/GutenbergKit/releases/tag/v0.11.1) is published and can now be integrated into this PR.

oguzkocer replied:
Updated in dc1e76f90147c7cc06f8f8d068bd73617fb267e8.

dcalhoun reviewed (commented):
I suggest we await a fix in a new release from GutenbergKit before merging this. 

Also, I noted what appears to be a crash in Chucker when quickly switching tabs, but this may be unrelated to our code. I share it nonetheless. 

<details><summary>Chucker crash on tab switch</summary>

https://github.com/user-attachments/assets/4ecb4a07-80c8-475c-83ee-d0c39c2d4b43

</details>

oguzkocer commented:
> Do we need to give the user access to this screen? Can we log it internally and access it any other way?

If we are using Chucker, we have to use its UI and there is no way to access the data as far as I can tell. We discussed the possibility of building something in-house with @jkmassel and the proposed `TrackNetworkRequestsInterceptor` design from this PR should make that switch relatively easy.

I have not been a big fan of using Chucker, but it does make sense to use that initially to test the flow and replace it with our implementation later. I think that's one reason Jeremy wanted to put this behind a feature flag as we currently don't expect end users to use this feature.

oguzkocer committed:
Update GutenbergKit to v0.11.1

oguzkocer commented:
> @oguzkocer I note the GutenbergKit network activity now displays in Chucker, but the timestamp and duration appear incorrect. Is the data provided by GutenbergKit incompatible or empty?

@dcalhoun This is a known limitation of the replay approach. The `ReplayInterceptor` returns the pre-recorded response immediately without making a real network call, so:

- **Timestamp**: Chucker records when the replay happens, not when the original JS fetch executed
- **Duration**: Shows ~0ms since the replay returns instantly (response is already available)

Chucker doesn't have an API to inject custom timestamps - it always uses its own timing. We could simulate duration by adding a wait time to the replay, but that adds overhead without much benefit. If we switch to an in-house tracking implementation in the future, we'd be able to record these values correctly.
